File: src/main/java/fi/quanfoxes/parser/patterns/SingletonPattern.java
package fi.quanfoxes.parser.patterns;

import java.util.List;

import fi.quanfoxes.lexer.Token;
import fi.quanfoxes.lexer.TokenType;
import fi.quanfoxes.parser.Context;
import fi.quanfoxes.parser.Node;
import fi.quanfoxes.parser.Pattern;
import fi.quanfoxes.parser.Singleton;

public class SingletonPattern extends Pattern {
    public static final int PRIORITY = 1;

    public SingletonPattern() {
        // Pattern:
        // Identifier / Function
        super(TokenType.IDENTIFIER | TokenType.FUNCTION | TokenType.NUMBER);
    }

    @Override
    public int priority(List<Token> tokens) {
        return PRIORITY;
    }

    @Override
    public boolean passes(List<Token> tokens) {
        return true;
    }

	@Override
	public Node build(Context context, List<Token> tokens) throws Exception {
		return Singleton.parse(context, tokens.get(0));
	}
}

File: Zigzag/Parser/Patterns/SingletonPattern.cs
using System.Collections.Generic;

public class SingletonPattern : Pattern
{
	public const int PRIORITY = 1;

	// Pattern:
	// Identifier / Function
	public SingletonPattern() : base(TokenType.IDENTIFIER | TokenType.FUNCTION | TokenType.NUMBER | TokenType.STRING | TokenType.CONTENT) { }

	public override int GetPriority(List<Token> tokens)
	{
		return PRIORITY;
	}

	public override bool Passes(Context context, List<Token> tokens)
	{
		return true;
	}

	public override Node Build(Context context, List<Token> tokens)
	{
		return Singleton.Parse(context, tokens[0]);
	}
}